src/tasks/WasmAppBuilder/coreclr/ManagedToNativeGenerator.cs

```diff
@@ -0,0 +1,225 @@
+
+            var resolver = new PathAssemblyResolver(managedAssemblies);
+            using var mlc = new MetadataLoadContext(resolver, "System.Private.CoreLib");
+            foreach (string asmPath in managedAssemblies)
+            {
+                log.LogMessage(MessageImportance.Low, $"Loading {asmPath} to scan for pinvokes, and icalls");
+                Assembly asm = mlc.LoadFromAssemblyPath(asmPath);
+                pinvoke.ScanAssembly(asm);
+                icall.ScanAssembly(asm);
+            }
+
+
+            IEnumerable<string> cookies = Enumerable.Concat(
+                pinvoke.Generate(PInvokeModules, PInvokeOutputPath),
+                Enumerable.Concat(icall.Generate(IcallOutputPath),
+                missingCookies));
+
+            var m2n = new InterpToNativeGenerator(log);
+            m2n.Generate(cookies, InterpToNativeOutputPath);
+
+            if (!string.IsNullOrEmpty(CacheFilePath))
+                File.WriteAllLines(CacheFilePath, PInvokeModules);
+        }
+
+        List<string> fileWritesList = new() { PInvokeOutputPath, InterpToNativeOutputPath };
+        if (IcallOutputPath != null)
+            fileWritesList.Add(IcallOutputPath);
+        if (!string.IsNullOrEmpty(CacheFilePath))
+            fileWritesList.Add(CacheFilePath);
+
+        FileWrites = fileWritesList.ToArray();
+
+        string FixupSymbolName(string name)
+        {
+            if (_symbolNameFixups.TryGetValue(name, out string? fixedName))
+                return fixedName;
+
+            fixedName = Utils.FixupSymbolName(name);
+            _symbolNameFixups[name] = fixedName;
+            return fixedName;
+        }
+
+    }
+
+    private bool ShouldRun(IList<string> managedAssemblies)
+    {
+        if (string.IsNullOrEmpty(CacheFilePath) || !File.Exists(CacheFilePath))
+        {
+            Log.LogMessage(MessageImportance.Low, $"Running because no cache file found at '{CacheFilePath}'.");
+            return true;
+        }
+
+        string oldModules = string.Join(",", File.ReadLines(CacheFilePath).OrderBy(l => l));
+        string newModules = string.Join(",", PInvokeModules.OrderBy(l => l));
+        if (!string.Equals(oldModules, newModules, StringComparison.InvariantCulture))
+        {
+            Log.LogMessage(MessageImportance.Low, $"Running because the list of pinvoke modules has changed from: {oldModules} to {newModules} .");
+            return true;
+        }
+
+        // compare against the output files
+        // get the timestamp for oldest output file
+        DateTime oldestOutputDt = DateTime.MaxValue;
+        if (CheckShouldRunBecauseOfOutputFile(IcallOutputPath, ref oldestOutputDt) ||
+            CheckShouldRunBecauseOfOutputFile(PInvokeOutputPath, ref oldestOutputDt) ||
+            CheckShouldRunBecauseOfOutputFile(InterpToNativeOutputPath, ref oldestOutputDt))
+        {
+            return true;
+        }
```

Comment:
We should be relying on the build system to handle incremental logic. Tasks should "just do it". Trying to manually handle incrementality is more confusing than helpful, I think.